src/urllib3/response.py

```diff
@@ -785,12 +763,86 @@ def read(
             self._fp_bytes_read += len(data)
             if self.length_remaining is not None:
                 self.length_remaining -= len(data)
+        return data
+
+    def _popleft_decoded_bytes(self, n: int) -> bytes:
+        decoded_data = bytes(self._decoded_bytes[:n])
+        del self._decoded_bytes[:n]
+        return decoded_data
+
+    def _popall_decoded_bytes(self) -> bytes:
+        decoded_data = bytes(self._decoded_bytes)
+        del self._decoded_bytes[:]
+        return decoded_data
 
-            data = self._decode(data, decode_content, flush_decoder)
+    def _decoded_bytes_in_buffer(self) -> int:
+        return len(self._decoded_bytes)
 
+    def read(
+        self,
+        amt: Optional[int] = None,
+        decode_content: Optional[bool] = None,
+        cache_content: bool = False,
+    ) -> bytes:
+        """
+        Similar to :meth:`http.client.HTTPResponse.read`, but with two additional
+        parameters: ``decode_content`` and ``cache_content``.
+
+        :param amt:
+            How much of the content to read. If specified, caching is skipped
+            because it doesn't make sense to cache partial content as the full
+            response.
+
+        :param decode_content:
+            If True, will attempt to decode the body based on the
+            'content-encoding' header.
+
+        :param cache_content:
+            If True, will save the returned data such that the same result is
+            returned despite of the state of the underlying file object. This
+            is useful if you want the ``.data`` property to continue working
+            after having ``.read()`` the file object. (Overridden if ``amt`` is
+            set.)
+        """
+        if amt is not None:
+            cache_content = False
+
+        if decode_content is None:
+            decode_content = self.decode_content
+
+        if not decode_content:
+            data = self.raw_read(amt)
             if cache_content:
                 self._body = data
+            return data
+
+        if amt is not None and self._decoded_bytes_in_buffer() >= amt:
+            return self._popleft_decoded_bytes(amt)
+
+        data = self.raw_read(amt)
+
+        if not data and self._decoded_bytes_in_buffer() == 0:
```

Comment:
Great question. I think raw_read() can only return 0 bytes when we reach EOF OR when amt is set to 0. My understanding is that in both cases it's OK return `b''` as well. Should we add a comment?

Reply:
How would you do that?

Reply:
Since `raw_read` doesn't modify the buffer, can we assume that `self._decoded_bytes_in_buffer()` will return the same value before and after the `raw_read()`? If so we can cache the value from before the `raw_read` call. Since this is getting into super-nit territory I'll approve the PR though :)

Reply:
No problem regarding nits, I still have the memory issue to investigate anyways.

But it's just a len() call and a function call so I'm not sure that there's a lot to gain with caching. Maybe I should just replace this function with `len(self._decoded_bytes)`?

Reply:
Yeah let's do that!